Try "constrained" content width as new layout type

tellthemachines opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Fixes #33374.

## New layout type

* Moves the content width logic to a new layout type, called **"constrained"** which makes it easier to set blocks to use content width by default.

The idea is that, when creating a new block type, we should only have to add
```
"__experimentalLayout": {
	"default": {
		"type": "constrained"
	}
}
```
to its `block.json` and the block will apply content width to its children by default. 

* Add the global content width styles as "base styles" for "constrained" layout, so we don't output `wp-container` styles for them (this is necessary so the default state of the column layout can access those styles).


## Changes to Group block and post editor layout

* Changes the Group block so **newly-added blocks** are automatically set to "constrained" layout. Note: for back-compat reasons, we couldn't just change the default layout type, so we're adding some logic to the block Edit to set the layout attribute for all new blocks.

* Adds deprecation/migration so older group blocks with content width can move to using the constrained layout.

* The post editor appearance and alignment settings for top-level blocks used to depend on the "default" layout config; this is now changed to use "constrained" layout as the full/wide width options are no longer available for "default".

## Changes to the sidebar Layout controls

* The layout "full width" toggle will now switch from this new layout type to the default "flow" layout.

* Changes the wording on the toggle to "Inner blocks respect content width" so that the default (flow) layout state is off.

* The custom content width options now appear when content width is enabled via the toggle.

## Things to consider

* Do we also want to change the layout type for Post Content and Column blocks? That would require using the same logic in their Edit functions as we're applying to Group here.
~~* Is "column" really the best name for this layout type? Suggestions most welcome!~~

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

#33374 has shown that for most container blocks in most cases, _not_ having a content width by default is unexpected and puts the burden on the theme author to manually change the setting every time. 

This changes the most used container block (Group) to have content width by default, and provides an easy way of enabling content width by default for new blocks.


## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

## Using a block theme:

1. Create a bunch of Group blocks with various layout settings on trunk and save.
2. Switch to this branch; check that previous blocks still work as expected.
3. Add a new Group block to a post or page.
4. Check that content width is set on its children by default.

## Using a classic theme:

1. Check that alignments on Group blocks and their alignment-enabled descendants (such as Image blocks) are working the same as on trunk.
2. Add a `theme.json` to the classic theme.
3. Check that newly added Group blocks have content width set by default, and their alignment-enabled descendants have full/wide alignment options.


## Screenshots or screencast <!-- if applicable -->

tellthemachines committed:
Try content width column as new layout type

andrewserong commented on lib/block-supports/layout.php:
```diff
@@ -43,6 +43,16 @@ function gutenberg_get_layout_style( $selector, $layout, $has_block_gap_support
 
 	$style = '';
 	if ( 'default' === $layout_type ) {
```
What happens to existing Group blocks with the default layout type, that set `contentSize` or `wideSize`?

andrewserong reviewed (commented):
Fascinating idea @tellthemachines!

To make sure I'm following along, the idea with the `default` layout then, is that it's a simpler flow-based container that can group blocks and provide vertical spacing and alignments, but doesn't provide content size settings? Thinking about the available layout types, I find the idea of `default/flow` being quite simple and leaving things like content size to other layout types quite compelling!

I left a comment on `layout.php`, but I'm wondering what happens to Group blocks that are out in the wild (particularly older post/page content that is updated infrequently) that currently set content / wide size or set `inherit` to `true`. If we need duplicate handling for those cases, then I could imagine things starting to get messy again, where the shape of this PR is looking quite neat so far 🤔

In terms of naming, I think back when the Stack group block variation was added, folks avoided using Column as the name, to try to avoid confusion with the Column block. That said, I quite like the idea of a Column layout type, particularly if we eventually support settings like flex-basis for max-width, or effectively the kinds of features that are used in the Column block!

tellthemachines commented:
> To make sure I'm following along, the idea with the default layout then, is that it's a simpler flow-based container that can group blocks and provide vertical spacing and alignments, but doesn't provide content size settings? Thinking about the available layout types, I find the idea of default/flow being quite simple and leaving things like content size to other layout types quite compelling!

Yes, that's the idea!

>  I'm wondering what happens to Group blocks that are out in the wild (particularly older post/page content that is updated infrequently) that currently set content / wide size or set inherit to true.

We should be able to handle those with a deprecation, migrating blocks with inherit to the column layout type. I'm not sure what the approach should be for third party blocks using layout though. Are there even any of those? And is it ok to have a breaking change if the layout support is still experimental?

tellthemachines committed:
Consolidate root selector rules.

andrewserong commented:
> Yes, that's the idea!

Thanks for confirming!

> We should be able to handle those with a deprecation, migrating blocks with inherit to the column layout type.

Unfortunately, I don't think a deprecation will be enough in this case as it requires the blocks to be edited in the editor before the rendering fix will take effect. This could be a tricky situation where we need non-migrated markup to still be rendered correctly 🤔

> I'm not sure what the approach should be for third party blocks using layout though. Are there even any of those?

I'm not sure, but the layout support doesn't appear to be intended for third party blocks at this stage, as it's pretty much undocumented?

> And is it ok to have a breaking change if the layout support is still experimental?

That's a great question. From my perspective, I think breaking changes in API-terms are probably fine, but breaking styles on existing sites, or requiring block markup to be migrated in order to render correctly is probably a blocker.

It's definitely one of the harder areas for me conceptually with the Layout block support — because it's used in production on such heavily used blocks (like Group and Columns) it's very difficult to make significant changes without breaking _something_ 😅

andrewserong commented:
> Unfortunately, I don't think a deprecation will be enough in this case as it requires the blocks to be edited in the editor before the rendering fix will take effect

Hrm... unless if we infer the type based on the presence of the `inherit` or `contentSize`, `wideSize` attributes at the top of the `gutenberg_get_layout_style` function... okay, I think I'm back on board 😅

tellthemachines commented:
Yeah, I was thinking the styles to display content width only need the `is-layout-column` class, which we could attach on the server side if it doesn't exist. I'll have a play and see what we can do!

tellthemachines committed:
Undo rebase weirdness

tellthemachines committed:
Add deprecation and server-side legacy support

tellthemachines commented:
I updated with a deprecation for the Group block, as well as some server-side logic to handle cases where there are `inherit` or `contentSize` values. There remains a problem though: 

* If an existing Group block _doesn't_ have any of those values set, it is assumed to be [using the `$default_block_layout`](https://github.com/WordPress/gutenberg/blob/try/new-layout-type/lib/block-supports/layout.php#L193). 
* Because we've changed the default for Group to the new 'column' layout, existing Group blocks with no particular layout configured will automatically start using 'column', which we don't want. We want them to continue using 'default'. 
* This will affect existing content on the front end.

So you did have a good point above @andrewserong 😅 

I'm not sure how best to fix this. We could, of course, not change the default on the Group block and instead do something like I tried in #42582 to update new blocks on the fly. I still don't love that solution though 😅

andrewserong commented:
> Because we've changed the default for Group to the new 'column' layout, existing Group blocks with no particular layout configured will automatically start using 'column', which we don't want. We want them to continue using 'default'.

Ah, yes, that's a good point. Thanks for digging in here! At the very least, it's really helpful to have these PRs that demonstrate a potential solution and the exact problems we encounter when attempting to implement it. It sounds like we've got a few trade-offs to weigh up here!

> I'm not sure how best to fix this.

In some ways it feels like one of the challenges we face is a similar one to some other issues/PRs of wanting to be able to perform an action when a block is inserted, and only on insert. I wonder if there's any way for us to tease apart the difference between the default value (in terms of what's a default value when the block is serialised) from the value that is set when a block is inserted? That might be a bit of an API rabbithole to figure out, though 🤔

From my perspective, between the approaches, it feels like #42582 gets to the heart of the particular task at hand, but like you mention, that solution has its drawbacks too.

I wonder if there's anything we can do with block variations, or use update logic that only applies if the Group block is empty? At the very least, that might help us hone in on only making changes to blocks that aren't doing anything visually on the front end of the site?

Not sure how helpful that is as an idea, but I'm happy to do a bit of digging / hacking around this week, too!

andrewserong commented:
> or use update logic that only applies if the Group block is empty?

Not sure how helpful it is, but I had a go at this in #42876 @tellthemachines — it's still a pretty hacky approach, but tries to see if we can get it kinda working without deprecations. I was mostly curious to see if restricting the logic to only apply to empty Group blocks that haven't yet defined `inherit` as either `true` or `false` helps us at all 🤔

tellthemachines commented:
Thanks for helping me think this through @andrewserong !

> I wonder if there's any way for us to tease apart the difference between the default value (in terms of what's a default value when the block is serialised) from the value that is set when a block is inserted?

I don't think there is at the point that matters to us here, which is in `layout.php` where we're [setting the value of `$used_layout`](https://github.com/WordPress/gutenberg/blob/try/new-layout-type/lib/block-supports/layout.php#L193).

* If the block has a non-default value for its layout type (such as `inherit` for the current definition of flow layout), its attributes reflect that. If the layout values on the block were never changed, we don't have any attributes at this point.
* If there are no attributes, we grab the default layout for that block type. This is where we run into a problem: if we change the default for the Group block, any existing block without attributes will now think it's meant to have a column layout.

I think attributes are our best bet to tell apart blocks with the legacy default layout from newly-added blocks that are meant to have the column layout by default, because they are the only info we have about each specific block. Unless I'm missing something, everything else we can access here - block defaults, layout definitions - concerns the block type or the layout type, so gives us nothing on the block itself.

The thing is we can't add attributes to existing blocks unless they are being edited and can go through the deprecation, but we _can_ add attributes to all new blocks as they are generated, which is what #42582 attempts to do. In which case, we don't need to change the block layout type, just change the edit logic to switch all new blocks to the column type as they're created.

> I wonder if there's anything we can do with block variations, or use update logic that only applies if the Group block is empty? At the very least, that might help us hone in on only making changes to blocks that aren't doing anything visually on the front end of the site?

I don't _think_ adding a new variation would help us here, because we'd still face the problem of old default vs new default. And looking at Group content is risky because an empty Group might be used for formal purposes, as in [this artwork](https://block-museum.com/2022/01/11/block-composition/)!

The more I think of it, the more I'm convinced that we need to do something like #42582. But I think what we're doing in this PR is also important, because it provides a way of configuring new block types to have content width from scratch. 

However, the logic from #42582 will need to change somewhat to accommodate this PR. I think I'll try bringing it in here, thus ending up with the mega-PR I was trying to avoid 😅

andrewserong commented:
> And looking at Group content is risky because an empty Group might be used for formal purposes, as in [this artwork](https://block-museum.com/2022/01/11/block-composition/)!

Hey, that's also a fantastic pattern to use as a test case to make sure we don't break anything! What a fun one 😍

And I see what you mean, from that pattern the following still renders on the frontend:

```
<!-- wp:group {"style":{"border":{"style":"solid"},"color":{"background":"#d3dce4"}}} -->
<div class="wp-block-group has-background" style="border-style:solid;background-color:#d3dce4"></div>
<!-- /wp:group -->
```

This is output on the site frontend as: 

<img width="690" alt="image" src="https://user-images.githubusercontent.com/14988353/182287786-87c2d158-7abb-4fa0-b7ae-ed03e4f331bb.png">

If we were going with an empty / has no inner blocks check, then it'd probably also need to check that there are no attributes set at all 🤔... ah, it sure is tricky! 😅

> I think attributes are our best bet to tell apart blocks with the legacy default layout from newly-added blocks that are meant to have the column layout by default

So, basically, yes, I think you've hit the nail on the head here 👍

tellthemachines committed:
Change group block approach

tellthemachines committed:
Fix buggy align behaviour

tellthemachines committed:
Alignments should work for classic themes

tellthemachines committed:
Set output to empty string.

tellthemachines committed:
Add layout type fallback for alignments

tellthemachines committed:
Regenerate fixtures

tellthemachines committed:
Update snapshots

tellthemachines committed:
Change nested blocks to buttons

tellthemachines committed:
Fix php test diffs

tellthemachines committed:
Fix rebase inconsistency in layout

tellthemachines committed:
Add test for content size variable output

tellthemachines committed:
Add test for column layout.

tellthemachines commented on lib/block-supports/layout.php:
```diff
@@ -42,6 +42,31 @@ function gutenberg_get_layout_style( $selector, $layout, $has_block_gap_support
+				);
+			}
+		}
+	} elseif ( 'column' === $layout_type ) {
```
This is a very weird diff but essentially, what I did here was move all the contentSize/wideSize logic into a new "column" layout type

tellthemachines commented on lib/block-supports/layout.php:
```diff
@@ -238,19 +263,23 @@ function gutenberg_render_layout_support_flag( $block_content, $block ) {
+	// Set the correct layout type for blocks using legacy content width.
+	if ( isset( $used_layout['inherit'] ) && $used_layout['inherit'] || isset( $used_layout['contentSize'] ) && $used_layout['contentSize'] ) {
+		$used_layout['type'] = 'column';
+	}
```
Blocks with `contentSize` or `inherit` are migrated to "column" layout on the front end.

tellthemachines commented on lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php:
```diff
@@ -770,7 +758,30 @@ function( $pseudo_selector ) use ( $selector ) {
+				$wide_size    = isset( $settings['layout']['wideSize'] ) ? $settings['layout']['wideSize'] : $settings['layout']['contentSize'];
+				$wide_size    = static::is_safe_css_declaration( 'max-width', $wide_size ) ? $wide_size : 'initial';
+				$block_rules .= '--wp--style--global--content-size: ' . $content_size . ';';
+				$block_rules .= '--wp--style--global--wide-size: ' . $wide_size . ';';
```
`contentSize` and `wideSize` values are added to root as custom properties. This is necessary so we can use them in the base styles declared in `theme.json`.

tellthemachines commented on packages/block-editor/src/components/block-alignment-control/use-available-alignments.js:
```diff
@@ -46,7 +46,7 @@ export default function useAvailableAlignments( controls = DEFAULT_CONTROLS ) {
 
 	// Starting here, it's the fallback for themes not supporting the layout config.
-	if ( layoutType.name !== 'default' ) {
+	if ( layoutType.name !== 'default' && layoutType.name !== 'column' ) {
```
We should in theory have been able to change this to `if ( layoutType.name !== 'column' )`, but that caused the `BlockAlignmentUI` unit tests to fail, because rendering that component is dependent on a layout type (which defaults to `default`) and in order for tests to succeed it needs to have some alignment controls inside it 😅

tellthemachines commented on packages/edit-post/src/components/visual-editor/index.js:
```diff
@@ -181,10 +181,12 @@ export default function VisualEditor( { styles } ) {
-
-		return undefined;
+		// Set column layout for classic themes so all alignments are supported.
+		return { type: 'column' };
```
We need this type set explicitly here so that the top-level blocks in the post editor always get full/wide alignment controls. 

Ideally, for block themes those controls would only appear if PostContent block had content width, but that's not how it currently works. I opened #42911 to track that issue.

andrewserong replied:
On line 248, we pass in the `defaultLayout` object to the `LayoutStyle` component, which is responsible for outputting the default layout rules for the root container and post title wrapper. It looks like we might need to add or pass in the `type: 'column'` here, too?

In testing it appears that with this PR the alignment rules aren't being applied to the post title wrapper (I'm testing with emptytheme):

| Trunk | This PR |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/14988353/183344273-2b9efc62-d90e-4297-a635-c5be380c33f9.png) | ![image](https://user-images.githubusercontent.com/14988353/183344285-4c39f757-32e7-4f57-8139-844524f9c896.png) |

tellthemachines reviewed (commented):
This is now ready for review! Don't be discouraged by the big diff, most of it is snapshot updates 😄 

This PR does two main things:

* Introduces a new layout type to deal with content/wide widths and their associated styles.
* Sets some logic in the Group block so newly-added Groups get content width by default.

These two things could be done in separate PRs, but this way we can get a complete picture of how this works as a fix for #33374.

tellthemachines commented on test/e2e/specs/editor/various/copy-cut-paste.spec.js:
```diff
@@ -188,11 +188,11 @@ test.describe( 'Copy/cut/paste', () => {
-					name: 'core/paragraph',
-					attributes: { content: 'P' },
+					name: 'core/button',
+					attributes: { text: 'Click' },
```
So this is an interesting one. Ostensibly these tests are checking whether cut/paste actions work correctly for nested blocks. But they are also checking if the blocks are [replaced more than once](https://github.com/WordPress/gutenberg/pull/34430) during paste. This last part started failing with this changeset, probably because the Group block now updates itself to use the column layout. So to preserve the original intent of the check, I've replaced the Group/Paragraph with Buttons/Button.

andrewserong reviewed (commented):
I love what you're doing with this PR @tellthemachines! The ah-ha moment for me was looking at the logic in the UI for editing what used to be the inherit setting in the group block.

With this PR applied the logic now appears to be:

* Toggle `Inner blocks respect content width` off results in `default` layout being applied
* Toggle `Inner blocks respect content width` to `true` and don't enter any values for content/wide width, and the root/default values are used, and no container classname is added
* Toggle `Inner blocks respect content width` to `true` and enter values for contentSize, and a container class is added with the desired max-width on the children:

<img width="1239" alt="image" src="https://user-images.githubusercontent.com/14988353/183345105-50f930f4-7bc9-47b4-9524-aca46f0439dc.png">

Overall this seems like a really nice way to remove the confusing idea of inheriting from the default layout, kudos for digging into how to get us from the current state of things to a more ideal representation of this kind of state! 👍

---

For the questions:

> Do we also want to change the layout type for Post Content and Column blocks? That would require using the same logic in their Edit functions as we're applying to Group here.

If we don't make changes to these blocks in this PR, will this PR introduce regressions for existing blocks out in the wild? Or to put it differently, is this PR a breaking change for all blocks that have opted in to the default layout type? 🤔

We might need to put together some test markup to take a look at the impact of the changes on those blocks.

> Is "column" really the best name for this layout type? Suggestions most welcome!

I can't quite think of a good alternative, but I think it'd be great if we can think of something that helps distinguish it from the Column block as I imagine it could be confusing the two using the same name. I suppose one question for the naming of the type is: are there additional features we'd like this layout type to have, that could help clarify its intended usage? I can't quite think of anything that isn't super clunky, like "flow with content sizes" or "constrain child content sizes"... ah naming is hard 😅

---

This PR changes a fair bit, so it might take me a little while to review in detail. I'll add comments here over the next couple of days, with things I find during testing, if you don't mind! I've added a comment about the post title wrapper that I noticed immediately.

ZebulanStanphill commented:
Not sure if it's a bug in `trunk` or not, but when you're using the standard "Group" variation and you press the "Stack" variation button, it actually switches to the "Row" variation the first time you press it.

Also, I'm really not comfortable with "column" being the right name for this new layout type, especially since it's actually the "default" layout that makes the Group behave most like the existing block literally called "Column". (Speaking of which, I wonder if in the future we could turn Column blocks into just a variation of the Group block.)

On the other hand, I see how the name is also somewhat fitting, since child blocks are constrained within a set (or inherited) content width... a "column", basically. Unfortunately, I don't have a better suggestion.

tellthemachines commented:
Thanks for the feedback @andrewserong  and @ZebulanStanphill !

> If we don't make changes to these blocks in this PR, will this PR introduce regressions for existing blocks out in the wild? Or to put it differently, is this PR a breaking change for all blocks that have opted in to the default layout type?

It won't introduce regressions on the front end, because we're adding logic to migrate blocks with "inherit" to the new layout type. But I just realised that if a block is set to "inherit", that's not being recognised in the editor 😬 so I think we'll need to add deprecation/migrations for those blocks if nothing else.

> Not sure if it's a bug in trunk or not, but when you're using the standard "Group" variation and you press the "Stack" variation button, it actually switches to the "Row" variation the first time you press it.

Oh weird, I don't think that's in trunk. It must be coming from my changes; I'll investigate.

> Also, I'm really not comfortable with "column" being the right name for this new layout type

I agree, it's a terrible name 😅 . I'm thinking "center" might work better, because it's a center aligned column of content? Still welcoming any ideas!

tellthemachines commented:
> I think we'll need to add deprecation/migrations for those blocks if nothing else.

Thinking about third-party blocks that might possibly be using layout supports: given it's still marked `__experimental`, it _should_ be ok to introduce a potentially breaking change? 

Checking in [WP Directory](https://wpdirectory.net/search/01G9EDP3HJT9G4MYAQBNNPD6SV), there's not a huge amount of use yet and most of the blocks using it seem to be on the `flex` layout.

tellthemachines committed:
Fix loss of layout attributes.

tellthemachines committed:
Make layout changes work for blocks without deprecation

tellthemachines committed:
Fix custom content size controls

tellthemachines commented:
Ok, I fixed the bug with group variations ✅  

I also think I found a way of avoiding block deprecations for blocks we don't want to change the defaults on. [4fe6cea](https://github.com/WordPress/gutenberg/pull/42763/commits/4fe6ceae7ace2ab0d68d35f6093db23047943f4f) and [da31904](https://github.com/WordPress/gutenberg/pull/42763/commits/da3190411d0e0066c0503a074e5798e3f81af252) add some changes to the layout hook that will ensure controls display correctly and necessary classnames are added to any blocks with legacy attributes, without having to change the actual blocks. This means that in practice, in the editor, a block with `inherit: true` will behave as if its layout type were `column` (or whatever we decide to call it), but its own attributes won't be modified.

This has the advantage of ensuring that any third party blocks using the legacy settings out in the wild will still work as expected.

I'd of course be hugely grateful for any further testing of these changes to make sure they work well in all scenarios 😄

eric-michel commented:
@tellthemachines thank you so much for working on this! I'm finally able to start transitioning our theme to the use of `theme.json` now that I know this is being addressed.

Regarding the layout options and names, I am still a bit confused by the absence of content width constraints being considered "default".

In our theme (pre `theme.json`) and in `Twenty Twenty-One`, content width constraints are applied to basically all containers (except for specialized containers like the `columns` block that can only contain one type of inner block). If I set a global content width constraint, I expect that constraint to be applied site-wide in all containers that contain normal content. Same for setting a `wide` alignment. That, to me, is the default.

Similarly, if I set `contentSize` and `wideSize` in `theme.json`, I expect those to be applied to all containers that contain normal content, just as I expect my `fontSizes` and `color.palette` settings to apply to all blocks.

But more flexibility is always good! So I think having more layout options is awesome. However, here's how I would organize them:

- **Default Layout** (what you're calling Column Layout)
  - is the default setting on all normal containers (`wp-block-group`, `wp-block-column`, `wp-block-cover__inner-container`, `wp-block-media-text__content`, etc)
  - applies `contentSize` and `wideSize` from `theme.json` to child blocks of these containers
- **Custom Layout**
  - reveals `Content` and `Wide` input controls to customize content and wide alignment max-widths
  - could potentially, in the future, make it so that set options could be added in `theme.json` to choose from rather than manually typing in values
- **Unconstrained Layout** or possibly **No Layout**
  - the current default on production, where all content is unconstrained by any `max-width`

So all the wall of text to say - I would not use the name "Column Layout" or "Center Layout" - I would use "Default Layout". That seems by far the most intuitive to me, because we're setting a layout in `theme.json`, so that feels like the "default". That will definitely be the most intuitive for my users who expect the contents of containers to behave the same as the top level blocks.

cbirdsong commented:
I don't hate "center", but it's sort of a weird fit if this is also going to handle `alignleft`/`alignright`?

tellthemachines commented:
Thanks for your comments @eric-michel and @cbirdsong !

> Regarding the layout options and names, I am still a bit confused by the absence of content width constraints being considered "default".

Our 'default' here means unstyled, like the default HTML layout before any CSS is added. This is important to keep because it's the state any block will default to if something goes wrong with its layout logic. I don't love the use of the word 'default' myself, but we don't want to create more back-compat trouble than strictly necessary with this change 😅 

> I don't hate "center", but it's sort of a weird fit if this is also going to handle alignleft/alignright?

That's a good point, also related to [this comment](https://github.com/WordPress/gutenberg/issues/38998#issuecomment-1209586749). I'm thinking it might be good to call it something less prescriptive like `constrain`.

tellthemachines commented:
Pinging @WordPress/gutenberg-design for some help with the naming conundrum!

jasmussen commented:
TwentyNineteen had a left aligned but constrained column, so I'd also avoid center. Even if "constrain" might be technical, I'd seek that avenue instead. "bounded"? There aren't any great terms, but will these be surfaced in the UI anywhere?

Speaking of, this is an impressive PR, but I get the sense that it's also touching some very delicate aspects of the layout engine, and plotting a new course for how themes are built. Mainly that means a high bar for testing and reviews, so we have as high as confidence as we can in the path forward, to reduce the risk we have to rewind a bit. Is there any chance this change could be marked as experimental in the first iteration?

tellthemachines commented:
> Is there any chance this change could be marked as experimental in the first iteration?

Technically it already is, because layout support hasn't been made stable yet 😅 
I don't think there's much more we can do from that perspective. On the plus side, [WP Directory](https://wpdirectory.net/search/01G9EDP3HJT9G4MYAQBNNPD6SV) tells us Layout isn't widely used by third party code yet, so extensive testing here should catch any major issues. But yes, careful review, manual and automated testing is needed! I added tests where I thought it made sense, and happy to add more if folks think it's necessary.

scruffian commented:
We have a pre-existing issue with group blocks whereby they don't wrap content which is aligned left/right within them, for example:
<img width="1440" alt="Screenshot 2022-08-11 at 11 53 06" src="https://user-images.githubusercontent.com/275961/184118345-7346f9ab-32c0-424f-8589-8f94b17b9a4d.png">

We can easily fix this by adding a clearfix to the parent, but this is risky as we might break existing blocks. If we're changing these layout styles would now be a good opportunity to add a clearfix?

scruffian commented:
This seems like a good idea to me!

eric-michel commented:
> Our 'default' here means unstyled, like the default HTML layout before any CSS is added. 

I'm going to push back on this again in a moment, but assuming this can't be changed, how about **Theme Layout**? This specifies that the layout selected is the one defined by the theme. It avoids semantic issues with words like "column" or "center".

> This is important to keep because it's the state any block will default to if something goes wrong with its layout logic. I don't love the use of the word 'default' myself, but we don't want to create more back-compat trouble than strictly necessary with this change 😅

I definitely understand the backward compatibility issues that could arise from this, and how painful that could be. I just want to make one more pitch here because this feels really important from a usability standpoint.

There are tons of times the word "default" is used to mean the defaults set in `theme.json`. If I select `default` for font size, I don't get the default HTML font size from my browser, I get the default set in `theme.json`. If I select `default` (i.e. don't select) a text color for a block, I don't get the default text color from my browser, I get the default set in `theme.json`. Same for background color, link color, font family, etc.

I recognize that layout is different from things like colors, font family, and font size, which can be applied to the `body` tag and then assumed to work globally. But _regular users_ don't know that. I'm a firm believer that the editor should create as easy and accessible an editing experience as possible, and that means consistency in nomenclature and sensible default behavior.

If "default" for font size, font family, and color means the values set by the theme in `theme.json`, it will make much more sense to users if "default" for layout means the standard content with options set in `theme.json` as well.

ZebulanStanphill commented:
@scruffian
>We have a pre-existing issue with group blocks whereby they don't wrap content which is aligned left/right within them, for example:

I think sometimes the lack of clearing is desirable for visual effect, so maybe this should be a toggle on the block? Not sure what the default should be, though.

@eric-michel I would prefer avoiding the term "default" for any of the layout options; it's too vague. Just call one "constrained" and the other "full width", "unconstrained", or something else.

tellthemachines committed:
Fix post title

tellthemachines committed:
Update new e2e snapshots

tellthemachines commented:
> There aren't any great terms, but will these be surfaced in the UI anywhere?

Just realised I forgot to answer this! The layout names don't appear in the UI, only as block classnames (`is-layout-[name]`) and as settings in block.json when defining a specific layout type for a block. Maybe this is why they've never had much love 😅 

> If "default" for font size, font family, and color means the values set by the theme in theme.json, it will make much more sense to users if "default" for layout means the standard content with options set in theme.json as well.

This is a good point, but the layout name won't actually be visible to users, only to extenders, so in that respect it won't be much of an issue.

I'll go with `constrained` as it seems the least bad of available options. I'd also prefer to avoid `default ` altogether, but it might be too late for that. In any case, this PR is just introducing a new type, so any other changes we might want to consider should be looked at separately.

> We can easily fix this by adding a clearfix to the parent, but this is risky as we might break existing blocks. If we're changing these layout styles would now be a good opportunity to add a clearfix?

@scruffian this PR doesn't change any actual styles, just moves the content size logic to a new layout type. The clearfix sounds like it should be its own PR, especially if it's a breaking change 😅

tellthemachines committed:
Change layout type name to "constrained"

jasmussen commented:
> Just realised I forgot to answer this! The layout names don't appear in the UI, only as block classnames (is-layout-[name]) and as settings in block.json when defining a specific layout type for a block. Maybe this is why they've never had much love

In that case, I would suggest it's okay to go with the more technical term that is the most descriptive. Should it be "constrain" instead of "constrained" though? If it goes against "flow" and "flex", wouldn't the ed at the end be equivalent to "flowing" and "flexing"? Just a morning thought.

eric-michel commented:
> This is a good point, but the layout name won't actually be visible to users, only to extenders, so in that respect it won't be much of an issue.

I totally misunderstood this! I think a more technical term is totally fine in this context. I was thinking of it from the user interface standpoint (assuming there would be options to choose from in the editor, and the naming would be exposed there). `Constrain` or `Constrained` seems great.

tellthemachines committed:
Update snapshots for new tests.

andrewserong reviewed (commented):
> Do we also want to change the layout type for Post Content and Column blocks? That would require using the same logic in their Edit functions as we're applying to Group here.

I did a bit of testing for these two blocks today. It looks like the server-rendering logic holds up nicely for existing markup, but I think we'd probably need to update the logic for these in the Edit too, in order for them to render correctly in the editor when opening older markup — the main case I encountered was when a custom content size is set.

I've uploaded some test markup that I used here in case it's helpful: https://github.com/andrewserong/gutenberg-test-markup/blob/main/layout/flow-layout-permutations.html

| Editor on `trunk` | Editor on this PR |
| --- | --- |
| <img width="1798" alt="image" src="https://user-images.githubusercontent.com/14988353/184823868-3414cb92-eb29-4dc4-8e15-19e35b311473.png"> | <img width="1769" alt="image" src="https://user-images.githubusercontent.com/14988353/184824487-23961873-61de-4457-b9d9-de183737db0b.png"> |

In the above screenshots, the right Column block is set to a custom content size for the first set of columns, and is set to inherit on the second set (which appears to be okay). Not pictured is with no content width settings set at all, as that appears to be the same on trunk and in this PR.

What do you think?

tellthemachines commented:
> Should it be "constrain" instead of "constrained" though? If it goes against "flow" and "flex", wouldn't the ed at the end be equivalent to "flowing" and "flexing"?

That is a very good question, and one I'm not sure how to best answer 😅 because both "flow" and "flex" could be either verb or noun.
I [initially suggested](https://github.com/WordPress/gutenberg/issues/38998#issuecomment-1211486040) "constrain", and some folks thought "constrained" worked better, so ended up going with that. I'm not super fussed about it though 🙂

glendaviesnz commented on lib/block-supports/layout.php:
```diff
@@ -237,19 +262,23 @@ function gutenberg_render_layout_support_flag( $block_content, $block ) {
+	$layout_classname       = '';
+	$use_global_padding     = gutenberg_get_global_settings( array( 'useRootPaddingAwareAlignments' ) ) && ( isset( $used_layout['inherit'] ) && $used_layout['inherit'] || isset( $used_layout['contentSize'] ) && $used_layout['contentSize'] );
+
 	if ( isset( $used_layout['inherit'] ) && $used_layout['inherit'] ) {
```
a nit, could this be moved up to just after the $used_layout declaration as it seems like the other declarations are not needed if this returns early?
```
	$global_layout_settings = gutenberg_get_global_settings( array( 'layout' ) );
	$default_block_layout    = _wp_array_get( $block_type->supports, array( '__experimentalLayout', 'default' ), array() );
	$used_layout                   = isset( $block['attrs']['layout'] ) ? $block['attrs']['layout'] : $default_block_layout;

	if ( isset( $used_layout['inherit'] ) && $used_layout['inherit'] ) {
		if ( ! $global_layout_settings ) {
			return $block_content;
		}
	}

```

tellthemachines replied:
Yeah that makes sense!

glendaviesnz commented on lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php:
```diff
@@ -793,7 +781,30 @@ function( $pseudo_selector ) use ( $selector ) {
+			* If there are content and wide widths in theme.json, output them
+			* as custom properties on the body element so all blocks can use them.
+			*/
+			if ( isset( $settings['layout']['contentSize'] ) && $settings['layout']['contentSize'] || isset( $settings['layout']['wideSize'] ) && $settings['layout']['wideSize'] ) {
```
doesn't `isset` check that an array key is both set and not null? just asking for my own learning as it seems like the second `&& $settings['layout']['contentSize']` check is not needed, but my PHP null/undefined checking knowledge may be lacking?

tellthemachines replied:
Oh, well spotted! It shouldn't be necessary, no.

glendaviesnz commented on packages/block-library/src/group/edit.js:
```diff
@@ -67,6 +70,16 @@ function GroupEdit( { attributes, setAttributes, clientId } ) {
+	const { __unstableMarkNextChangeAsNotPersistent } =
+		useDispatch( blockEditorStore );
+	const { type: layoutType = null } = layout;
+	useEffect( () => {
```
In my testing this seems to just set the layout type attribute to what it is already set to. The toggling between `constrained` and `default` seems to work as expected without this `useEffect`, but maybe there is an edge case it was added for that I am not covering.

tellthemachines replied:
Yeah, this is a weird one. Though the block is inserted with the attribute, it's not saved unless we explicitly set it inside the effect. So if you add a block, save and refresh the layout type won't be there anymore. I'm not sure why this happens though 😅

glendaviesnz replied:
hmm, yeh, this is weird 🤔

ZebulanStanphill replied:
Perhaps this is due to Gutenberg's behavior (as I recall) of not serializing a block attribute when it's value is equal to the default? Heading blocks, for example, don't save their level when their level is set to `2`, since that's the default.

tellthemachines replied:
Good point, that would explain why we need to explicitly set the attribute - we're probably checking for any `setAttribute` calls, instead of actually looking at the attributes in the markup and comparing them to defaults.

talldan replied:
@tellthemachines This seems to have caused a bug - https://github.com/WordPress/gutenberg/issues/43760.

I'm not sure exactly what's happening, but this effect seems to send `useBlockSync` into a frenzy in particular situations.

> we're probably checking for any setAttribute calls, instead of actually looking at the attributes in the markup and comparing them to defaults.

I think it's more that `layout` is an object type attribute, so will be referentially different to the default when set via `setAttributes`. The code that checks for default values probably only does a `===` comparison. For simple types setting to the default will result in the attribute not being output.

> So if you add a block, save and refresh the layout type won't be there anymore.

I think this is because layout code is using the 'parsed' block object, and that doesn't have defaults applied.

Defaults are applied when accessing the attributes via the `WP_Block` object. There's the magic getter here:
https://github.com/WordPress/wordpress-develop/blob/b80ba269f6cf0b0600be49b89b9a926421af88ee/src/wp-includes/class-wp-block.php#L164-L191

I think the layout code should switch to using the `WP_Block` if possible. If I'm looking at the right code, layout hooks into `render_block` and the `WP_Block` instance is available as the third param of the callback:
https://github.com/WordPress/wordpress-develop/blob/b80ba269f6cf0b0600be49b89b9a926421af88ee/src/wp-includes/class-wp-block.php#L284

tellthemachines committed:
Content size variables should be output in the site editor.

tellthemachines commented on packages/edit-site/src/components/global-styles/test/use-global-styles-output.js:
```diff
@@ -484,6 +484,19 @@ describe( 'global styles renderer', () => {
+				},
+			};
+			expect( toStyles( tree, 'body' ) ).toEqual(
+				'body {margin: 0; --wp--style--global--content-size: 840px; --wp--style--global--wide-size: 1100px;}.wp-site-blocks > .alignleft { float: left; margin-right: 2em; }.wp-site-blocks > .alignright { float: right; margin-left: 2em; }.wp-site-blocks > .aligncenter { justify-content: center; margin-left: auto; margin-right: auto; }'
```
I'm not sure why the `.wp-site-blocks` rules are output here because we don't seem to be using that classname in the site editor. Or am I missing something?

andrewserong replied:
It looks like the classname is added to the `BlockList` wrapper here in the site editor: https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-site/src/components/block-editor/index.js#L199

I'm still not 100% sure how it's all hooked together, though, I get a little lost following the editor's component hierarchy 😅

tellthemachines replied:
Oh right, I can see it now. For some reason it wasn't popping up on find in the inspector 😕 All good then!

tellthemachines committed:
Fix custom content sizes in editor.

tellthemachines committed:
Tidy up PHP.

tellthemachines commented:
Thanks for the feedback folks! I think I've addressed everything, plus an extra bug that became apparent when testing with Andrew's markup 😄 

> It looks like the server-rendering logic holds up nicely for existing markup, but I think we'd probably need to update the logic for these in the Edit too, in order for them to render correctly in the editor when opening older markup — the main case I encountered was when a custom content size is set.

I managed to fix this without changing anything in the actual blocks, so legacy markup for any existing block should just work now - which is a plus for back compat! especially if there are any custom blocks out there using the layout support.

glendaviesnz commented:
@tellthemachines are you planning to add a new deprecation fixture to match the new deprecation that has been added, eg. `/test/integration/fixtures/blocks/core__group__deprecated-3.html`? It maybe needs a couple adding to represent the different layout combinations that can be migrated. Could be a follow-up PR if you think this one is big enough already.

glendaviesnz commented:
It seems to be testing ok for me, but haven't had a chance to look at classic themes yet, will try and do that tomorrow.

tellthemachines committed:
Add deprecation fixtures

tellthemachines commented:
Good idea, @glendaviesnz - I added a deprecation fixture for the current default, plus a variation with the "inherit" layout and its updated version.

ramonjd committed:
Fixed duplicated code after rebase hack job. Sorry!
Fixed incoming failing test (formatted only).

ramonjd commented:
I've been testing with various configurations, and whatever I do, I can't manage to break it! 😄 

In a block theme, I added some group blocks in trunk (they had `is-layout-flow` by default):

<img width="784" alt="Screen Shot 2022-08-22 at 12 16 00 pm" src="https://user-images.githubusercontent.com/6458278/185828539-a155aae5-f46c-42f7-ab0d-f3095c0023c7.png">

Switched to this branch and I see that the containing block has the default layout of `is-layout-constrained`:

<img width="766" alt="Screen Shot 2022-08-22 at 12 17 06 pm" src="https://user-images.githubusercontent.com/6458278/185828527-45b980af-1f1f-4f21-9741-1f6120942bcc.png">

The inner blocks receive the following CSS:

```css
body .is-layout-constrained > :where(:not(.alignleft):not(.alignright):not(.alignfull)) {
    max-width: var(--wp--style--global--content-size);
    margin-left: auto !important;
    margin-right: auto !important;
}
```

New group block patterns have the `is-layout-constrained` by default:

<img width="515" alt="Screen Shot 2022-08-22 at 12 55 19 pm" src="https://user-images.githubusercontent.com/6458278/185829570-c5a6ac43-d5d8-434b-9124-3c19054f87b0.png">

Classic themes behave the same between this branch and trunk (tested 2011, 2014, 2020, 2021)

When I add a theme.json to 2021, I'm seeing `is-layout-constrained` on the site container and new group blocks/resaved group blocks are `is-layout-constrained` by default.

ramonjd reviewed (approved):
Mostly manual testing, but it LGTM
Nothing stands out code-wise, tests pass.

andrewserong commented on lib/block-supports/layout.php:
```diff
@@ -42,6 +42,31 @@ function gutenberg_get_layout_style( $selector, $layout, $has_block_gap_support
+						),
+					),
+					array(
+						'selector'     => "$selector > * + *",
```
The specificity was increased on the comparable selector recently to be `'selector'     => "$selector$selector > * + *",` to workaround how the style engine processor groups together rules. Shall we add that in here, too?

andrewserong replied:
Here's the PR where it was introduced: https://github.com/WordPress/gutenberg/pull/43052

tellthemachines committed:
Increase block gap selector specificity.

andrewserong reviewed (approved):
Great work here @tellthemachines, and thank you for all the follow-ups and iterations. I've just re-tested with the following markup saved to posts prior to switching to this branch, and the existing markup and permutations all appear to still be working well (both on the site frontend without applying any migrations, and when opened in the editor)

* [Group block permutations](https://github.com/andrewserong/gutenberg-test-markup/blob/main/layout/group-block-layout-permutations.html) 
* [Blocks with flow layout permutations](https://github.com/andrewserong/gutenberg-test-markup/blob/main/layout/flow-layout-permutations.html)

Since there seems to be reasonable consensus on the new layout type name, and we seem to be happy with the changes to the serialized post content, I think this looks good to merge, too! ✨

eric-michel commented:
Thank you for all your hard work on this @tellthemachines! This will finally allow us to adopt `theme.json` after a long wait.

youknowriad commented on packages/block-library/src/group/block.json:
```diff
@@ -15,6 +15,12 @@
 			"type": [ "string", "boolean" ],
 			"enum": [ "all", "insert", false ]
+		},
+		"layout": {
```
Why is this being set here since this attribute comes from the block support?

tellthemachines replied:
If we set the default in the block support, it would break back compat for existing blocks. Adding it as default here, together with the effect that's causing problems (#43760) in the edit function, ensures new blocks get the correct layout while existing blocks aren't changed.

If you have any ideas on how to do this better, I'm happy to explore them!

youknowriad commented:
> Changes the Group block so newly-added blocks are automatically set to "constrained" layout.

This seems like the main reason for this new layout right? and if I'm understanding right, this is the same thing as applying "inherit" by default. So why did we decide to introduce a new layout instead of just making "inherit" true by default?

tellthemachines commented:
> So why did we decide to introduce a new layout instead of just making "inherit" true by default?

Mainly to make it easier to set it as the default layout type on new blocks, but it also had the nice side effect of allowing us to make the content width styles base styles (with content and wide width output as custom properties at root level), so we don't have to output `wp-container-x` classes for them anymore. Strictly speaking it wasn't necessary to do this in order to make Groups have content width automatically, but it felt like a good improvement to make.

youknowriad commented:
> Strictly speaking it wasn't necessary to do this in order to make Groups have content width automatically, but it felt like a good improvement to make.

I might differ in opinion because I feel like constrained and flow are just the same thing with different configs and we're tweaking the concept of "layout" for convenience rather than actual semantics but anyway, seems a bit late to alter it even more.

> Adding it as default here, together with the effect that's causing problems (https://github.com/WordPress/gutenberg/issues/43760) in the edit function, ensures new blocks get the correct layout while existing blocks aren't changed.

I think it's just not possible, `useEffect` on blocks on initialization is something we should forbid IMO, it's not the first time it creates issues. One potential solution is to remove the "group" block from the inserter by default and instead use a variation with the layout set. `createBlock( 'core/group' )` will still result in a "flow" layout default.

tellthemachines commented:
> I think it's just not possible, useEffect on blocks on initialization is something we should forbid IMO, it's not the first time it creates issues.

I was afraid you might say that 😅 

>One potential solution is to remove the "group" block from the inserter by default and instead use a variation with the layout set. createBlock( 'core/group' ) will still result in a "flow" layout default.

That sounds like an acceptable compromise. Could we name the variation "Group" so that the creation flow doesn't change?

youknowriad commented:
> That sounds like an acceptable compromise. Could we name the variation "Group" so that the creation flow doesn't change?

Yes, I think so.
